test/ha/ha.go

```diff
@@ -23,12 +23,79 @@ import (
 	"time"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"knative.dev/pkg/test"
 	"knative.dev/pkg/test/logging"
 )
 
-// WaitForNewLeader waits until the holder of the given lease is different from the previousLeader.
+// GetLeaders collects all of the leader pods from the specified deployment.
+func GetLeaders(client *test.KubeClient, deploymentName, namespace string) ([]string, error) {
+	leases, err := client.Kube.CoordinationV1().Leases(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting leases for deployment %q: %w", deploymentName, err)
+	}
+	var pods []string
+	for _, lease := range leases.Items {
+		if lease.Spec.HolderIdentity == nil {
+			continue
+		}
+		pod := strings.Split(*lease.Spec.HolderIdentity, "_")[0]
+
+		// Deconstruct the pod name and look for the deployment.  This won't work for very long deployment names.
+		parts := strings.Split(pod, "-")
+		if len(parts) < 3 {
+			continue
+		}
+		if strings.Join(parts[:len(parts)-2], "-") != deploymentName {
+			continue
+		}
+		pods = append(pods, pod)
+	}
+	return pods, nil
+}
+
+// WaitForLeaders waits until there are "n" leases and returns a set with the names of the pods holding those leases.
+// Note that the "n" leases may be held by "m" < "n" pods, where "m" can be 1.
+func WaitForLeaders(client *test.KubeClient, deploymentName, namespace string, n int) (sets.String, error) {
+	span := logging.GetEmitableSpan(context.Background(), "WaitForLeaders/"+deploymentName)
+	defer span.End()
+
+	var leaders sets.String
+	err := wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		currentLeaders, err := GetLeaders(client, deploymentName, namespace)
+		if err != nil {
+			return false, err
+		}
+		if len(currentLeaders) < n {
+			return false, nil
+		}
+
+		leaders = sets.NewString(currentLeaders...)
+		return true, nil
+	})
+	return leaders, err
+}
+
+// WaitForNewLeaders waits until the collection of current leaders does not include the specified prior leaders.
+func WaitForNewLeaders(client *test.KubeClient, deploymentName, namespace string, previousLeaders sets.String) error {
+	span := logging.GetEmitableSpan(context.Background(), "WaitForNewLeaders/"+deploymentName)
+	defer span.End()
+
+	return wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		currentLeaders, err := GetLeaders(client, deploymentName, namespace)
+		if err != nil {
+			return false, err
+		}
+		if previousLeaders.HasAny(currentLeaders...) {
+			// TODO(mattmoor): Log overlap?
+			return false, nil
+		}
+		return true, nil
```

Comment:
`return !previousLeaders.HasAny(currentLeaders...), nil`

Reply:
I combined the methods to be closer to the style in use before, so this is a bit impractical now.